### ISSUE TITLE
No optimization after 'manually' adjusting the weights

### ISSUE THREAD

**Issue by SaremS** [rstudio/keras#105]
Hi,

I am currently trying to implement a Particle-Swarm - Backpropagation hybrid for training a standard (classification) feedforward NN in R (PSO first, then Backprop). Now, the problem that I keep running into is that once I have let PSO adjust the weights through set_weights and then start backpropagation, there won't be any increase in accuracy at all. I have tried through all the different keras optimizers and using only one iteration of PSO (local minimum was my first guess here) but as soon as I manually adjust the weights, backpropagation itself does not seem to work anymore, although PSO cleary optimizes the weights at first.

Do you have any idea how to fix this?

Thanks a lot,
Sarem

**Comment by jjallaire** [comment]
Regrettably this is far outside my expertise! Perhaps someone else will have an idea. One tip if you want to solicit help is to provide a working, reproducible example that illustrates the problem (that way others can try it out and tinker with it easily)

**Comment by topepo** [comment]
> once I have let PSO adjust the weights through set_weights and then start backpropagation, there won't be any increase in accuracy at all.

Well good job then! =]

But seriously, try manually setting the parameters to really bad values and see what happens. This should help us tease out whether it is an issue with the code/syntax or with the optimization problem. 

Also, try changing your optimizer to cross-entropy or something more sensitive than accuracy. We don't know much about the problem, but it is conceivable that an objective function based on hard predictions would be much easier to satisfy than something based on the class probabilities/soft predictions (at least that was what I learned from reading 10 years of boosting research).

**Comment by SaremS** [comment]
@topepo Thanks for the cross-entropy idea. What I found out was that once I let the PSO-part touch the keras network the class-probability for every sample will (always) be 1.0 for one class and 0.0 for every other. I am little confused at this point because I thought that the softmax-activation in the output layer (which I definitely included :D ) would make this impossible to happen.

**Comment by topepo** [comment]
> the class-probability for every sample will (always) be 1.0 for one class and 0.0 for every other. 

For which samples? I would expect that to happen for the training set data.

**Comment by SaremS** [comment]
Yes, I mean the training data. But why would class probability always be 0.0 or 1.0? As far as I know, the Softmax in the end maps the output in the final layer into (0;1) while the sum of the mapped output is 1 for each datapoint (so that we have probabilities). This is actually the case when I train the network with backpropagation only and then use _predict_proba()_

**Comment by topepo** [comment]
Even though the network might be fairly well regularized, it is usually very overfit to the training data (since these data were used to estimate the model). Try using the `keras` option to set aside a percentage of data for validation or use a resampling method to measure performance.

**Comment by SaremS** [comment]
Ok, I solved it. Apart from some minor bugs, letting PSO optimize for Cross-Entropy instead of Accuracy was a big help - thanks, @topepo 👍  So, there was no bug in keras at all.

**Comment by warproxxx** [comment]
@SaremS  would you mind sharing the code for PSO model in Keras?
